hystrix/metrics.go

```diff
@@ -46,85 +36,88 @@ func newMetrics(name string) *metrics {
+	if len(m.metricCollectors) < 1 {
+		panic("No Metric Collectors Registered.")
+	}
+	collection, ok := m.metricCollectors[0].(*metric_collector.DefaultMetricCollector)
+	if !ok {
+		panic("Default metric collector is not registered correctly. The default metric collector must be registered first.")
+	}
+	return collection
+}
+
+func (m *metricExchange) Monitor() {
 	for update := range m.Updates {
 		// we only grab a read lock to make sure Reset() isn't changing the numbers
 		// Increment() and Add() implement their own internal locking
 		m.Mutex.RLock()
 
-		// combined metrics
-		m.numRequests.Increment()
-		if update.Type != "success" {
-			m.Errors.Increment()
-		}
-
-		// granular metrics
-		if update.Type == "success" {
-			m.Successes.Increment()
-		}
-		if update.Type == "failure" {
-			m.Failures.Increment()
-		}
-		if update.Type == "rejected" {
-			m.Rejected.Increment()
-		}
-		if update.Type == "short-circuit" {
-			m.ShortCircuits.Increment()
-		}
-		if update.Type == "timeout" {
-			m.Timeouts.Increment()
-		}
-
-		// fallback metrics
-		if update.Type == "fallback-success" {
-			m.FallbackSuccesses.Increment()
-		}
-		if update.Type == "fallback-failure" {
-			m.FallbackFailures.Increment()
-		}
-
 		totalDuration := time.Now().Sub(update.Start)
-		m.TotalDuration.Add(totalDuration)
-		m.RunDuration.Add(update.RunDuration)
+		for _, collector := range m.metricCollectors {
+			collector.IncrementAttempts()
+			if update.Type != "success" {
+				collector.IncrementErrors()
+			}
+
+			// granular metrics
+			if update.Type == "success" {
+				collector.IncrementSuccesses()
+			}
+			if update.Type == "failure" {
+				collector.IncrementFailures()
+			}
+			if update.Type == "rejected" {
+				collector.IncrementRejects()
+			}
+			if update.Type == "short-circuit" {
+				collector.IncrementShortCircuits()
+			}
+			if update.Type == "timeout" {
+				collector.IncrementTimeouts()
+			}
+
+			// fallback metrics
+			if update.Type == "fallback-success" {
+				collector.IncrementFallbackSuccesses()
+			}
+			if update.Type == "fallback-failure" {
+				collector.IncrementFallbackFailures()
+			}
+
+			collector.UpdateTotalDuration(totalDuration)
+			collector.UpdateRunDuration(update.RunDuration)
+		}
 
 		m.Mutex.RUnlock()
 	}
 }
 
-func (m *metrics) Reset() {
+func (m *metricExchange) Reset() {
 	m.Mutex.Lock()
 	defer m.Mutex.Unlock()
 
-	m.numRequests = newRollingNumber()
-	m.Errors = newRollingNumber()
-
-	m.Successes = newRollingNumber()
-	m.Rejected = newRollingNumber()
-	m.ShortCircuits = newRollingNumber()
-	m.Failures = newRollingNumber()
-	m.Timeouts = newRollingNumber()
-
-	m.FallbackSuccesses = newRollingNumber()
-	m.FallbackFailures = newRollingNumber()
-
-	m.TotalDuration = newRollingTiming()
-	m.RunDuration = newRollingTiming()
+	for _, collector := range m.metricCollectors {
+		collector.Reset()
+	}
 }
 
-func (m *metrics) Requests() *rollingNumber {
+// This is not thread safe.
```

Comment:
I think this was a leftover note from when i was hacking earlier - Ill remove it. :D